libevmasm/RuleList.h

```diff
@@ -248,53 +254,53 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart5(
 	{
 		Word value = Word(1) << i;
 		rules.push_back({
-			{Pattern::Builtins::MOD, {X, value}},
-			[=]() -> Pattern { return {Pattern::Builtins::AND, {X, value - 1}}; },
+			Builtins::MOD(X, value),
+			[=]() -> Pattern { return Builtins::AND(X, value - 1); },
 			false
 		});
 	}
 
 	// Replace SHL >=256, X with 0
 	rules.push_back({
-		{Pattern::Builtins::SHL, {A, X}},
+		Builtins::SHL(A, X),
 		[=]() -> Pattern { return Word(0); },
 		true,
 		[=]() { return A.d() >= Pattern::WordSize; }
 	});
 
 	// Replace SHR >=256, X with 0
 	rules.push_back({
-		{Pattern::Builtins::SHR, {A, X}},
+		Builtins::SHR(A, X),
 		[=]() -> Pattern { return Word(0); },
 		true,
 		[=]() { return A.d() >= Pattern::WordSize; }
 	});
 
 	// Replace BYTE(A, X), A >= 32 with 0
 	rules.push_back({
-		{Pattern::Builtins::BYTE, {A, X}},
+		Builtins::BYTE(A, X),
 		[=]() -> Pattern { return Word(0); },
 		true,
 		[=]() { return A.d() >= Pattern::WordSize / 8; }
 	});
 
-	for (auto const& op: std::vector<Instruction>{
-		Pattern::Builtins::ADDRESS,
-		Pattern::Builtins::CALLER,
-		Pattern::Builtins::ORIGIN,
-		Pattern::Builtins::COINBASE
+	for (auto instr: {
```

Comment:
Won't ``for (auto const& instr: { Pattern::Builtins::ADDRESS, ... }) { ...Bultins::AND(instr(), mask)... }`` work?

Reply:
Ah no, those don't have the same types...

Reply:
Yeah, to be really nice this would require a compile time foreach on types, but that's probably overkill... especially since I'm not that much into pulling in ``boost::mpl`` for that...

Reply:
Haha, get this - I think this should actually work (needs ``#include <tuple>``):

```
std::apply([&](auto&&... instr) { (([&](auto&& instr) {
            assertThrow(Pattern::WorSize > 160, OptimizerException, "");
            Word const mask = (Word(1) << 160) - 1;
            rules.push_back({
                Bultins::AND(inst(), mask)
            }, false);
        })(instr), ...);},
        std::make_tuple(Pattern::Builtins::ADDRESS, Pattern::Builtins::CALLER);
```

Not that this is a reasonable thing to actually do :-). Maybe I'll find a nicer one...

Reply:
Ah, yes, it can be done rather nicely with one little helper like this:
```
template<typename C, typename... Args>
void foreach(C&& c, Args&&... args) { (c(std::forward<Args>(args)), ...) };
```

To be used like this:
```
foreach([](auto instr) {
  instr(); // iterates through "Pattern::Builtins::ADDRESS();" and "Pattern::Builtins::CALLER();", etc.
}, Pattern::Builtins::ADDRESS, Pattern::Builtins::CALLER);
```

Reply:
Nice - the same would have been a major pain without C++17 fold expressions.

Reply:
Maybe we can revisit if this does not work for wasm? :)